### PR TITLE
Feature/228 initial database structure

### PR DESCRIPTION
**Pull request by martincekada**
Task: [228](https://app.productive.io/1-infinum/dashboard/widgets/task/1187242)

#### Aim
Create initial database structure as it is defined in task

Specs are omitted on purpose (because of simplicity of models). Please comment if you think that there should be specs and what should they look like.

### CODE REVIEW

File: app/models/booking.rb [hunk 1]
```diff
@@ -0,0 +1,4 @@
+class Booking < ApplicationRecord
```
Comment (d4be4st) [inline app/models/booking.rb:1]: where is the annontation? :)

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+  has_many :bookings, dependent: :destroy
+
+  # discuss email format validations
```
Comment (d4be4st) [inline app/models/user.rb:4]: really good point!

File: db/migrate/20200910093434_create_flights.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+class CreateFlights < ActiveRecord::Migration[6.0]
+  def change
+    create_table :flights do |t|
+      t.belongs_to :company, null: false, foreign_key: { on_delete: :cascade }
+
+      t.string    :name
```
Comment (d4be4st) [inline db/migrate/20200910093434_create_flights.rb:6]: should we complicate It a bit and add airoplane model? 
maybe have an api which returns all airoplanes for a company? 

This would be extra, not sure if needed

**Review by d4be4st — CHANGES_REQUESTED** [review]

File: db/migrate/20200910093434_create_flights.rb (hunk 3 above)
Comment (martincekada) [reply db/migrate/20200910093434_create_flights.rb:6]: Noted this as an idea. Might be added as new feature request.

File: db/migrate/20200910093434_create_flights.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+      t.integer   :no_of_seats
+      t.integer   :base_price
+      t.timestamp :flys_at
```
Comment (lovro-bikic) [inline db/migrate/20200910093434_create_flights.rb:9]: A typo, correct form of the verb is `flies_at`.

**Review by cukoreva — APPROVED** [review]

File: db/migrate/20200910094305_create_bookings.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+class CreateBookings < ActiveRecord::Migration[6.0]
+  def change
+    create_table :bookings do |t|
+      t.belongs_to :flight, null: false, foreign_key: { on_delete: :cascade }
```
Comment (uncoverd) [inline db/migrate/20200910094305_create_bookings.rb:4]: I think @JureCindro has some strong opinions on why we need the cascade option and I mostly agree with him.

My gripe with using this is as follows:
If there was no other choice, and I was brave enough to go editing records in the DB directly, I would forget about this option and delete records I didn't want to accidentally. We're used to the concept of `dependent: :destroy` in Rails, but I don't think we carry that thought over to SQL.

Some articles (e.g. https://thoughtbot.com/blog/referential-integrity-with-foreign-keys) seem to even suggest using just cascades and ditching the callbacks. 

Does anyone else have any thoughts on this?

File: app/models/user.rb (hunk 2 above)
Comment (cilim) [reply app/models/user.rb:19]: do we maybe want to create a gem for this? It might be totally custom, eg. @JureCindro used this on nlbl and I stole it for Messe 😄 

```
class EmailValidator < ActiveModel::EachValidator
  # rubocop:disable Style/RedundantRegexpEscape
  EMAIL_REGEXP = %r{
    \A
    [a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@
    [a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?
    (?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+
    \z
  }x.freeze
  # rubocop:enable Style/RedundantRegexpEscape

  def validate_each(record, attribute, value)
    return if value =~ EMAIL_REGEXP

    record.errors[attribute] << options.fetch(:message, 'is not a valid email')
  end
end

`validates :email, email: true`
```

File: db/migrate/20200910092513_create_users.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+class CreateUsers < ActiveRecord::Migration[6.0]
```
Comment (cilim) [inline db/migrate/20200910092513_create_users.rb:1]: aren't we usually creating users with devise? Or is this whole example app going to be without authentication?

File: db/migrate/20200910093434_create_flights.rb (hunk 4 above)
Comment (cilim) [reply db/migrate/20200910093434_create_flights.rb:9]: `departs_at`
`arrives_at`

Sounds more intuitive, as it's the actual nomenclature used on airports.

File: db/migrate/20200910094305_create_bookings.rb (hunk 5 above)
Comment (cilim) [reply db/migrate/20200910094305_create_bookings.rb:4]: > Some articles even suggest using just cascades and ditching the callbacks.

Using `after_destroy` on the relationships that don't have a `dependent` option won't work then, as the records get destroyed on the db level...which might not be a bad thing considering all the callback hell arguments, but sometimes it really makes sense, so having this as a convention and use it as a must on all projects seems to strict.

My thoughts on the cascade option is that it's a bit scary if you go to the production rails console and delete stuff, so raising errors seem like the way to go, but I did find it annoying when I used the dashboard, went to delete a parent record that has relationships with foreign_key: true and got an error, which meant I'd have to delete all the relationships first manually and then the parent record.

So, all in all, I'd keep this flexible and add the cascade option on small relationships, that usually don't have their own relationships. For the "god models" I'd keep the foreign_key: true.

File: db/migrate/20200910094305_create_bookings.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+      t.belongs_to :user,   null: false, foreign_key: { on_delete: :cascade }
+
+      t.integer :no_of_seats
```
Comment (cilim) [inline db/migrate/20200910094305_create_bookings.rb:7]: maybe also think about complicating the seat logic a bit, so we have more cases to use.

We could have a `Seat` model maybe, which would belong to `Airplane` as @d4be4st suggested, and then go from there :)

**Review by cilim — APPROVED** [review]

File: db/migrate/20200910092513_create_users.rb (hunk 6 above)
Comment (d4be4st) [reply db/migrate/20200910092513_create_users.rb:1]: you can add devise later :D

File: app/models/user.rb (hunk 2 above)
Comment (d4be4st) [reply app/models/user.rb:19]: like https://github.com/afair/email_address ? :)

File: db/migrate/20200910094305_create_bookings.rb (hunk 5 above)
Comment (JureCindro) [reply db/migrate/20200910094305_create_bookings.rb:4]: Why would you differentiate rails relationship definitions from the ones defined on the db level? 

You should use AR's destroy method most of the times which uses the dependent option and destroys the related records, sql foreign key cascade wont run in those cases as AR destoys the records before the parent record is deleted.

But sometimes its requred to make a fast query that deletes a lot of records and all the associated records and unless you have cascade rules set that follow the dependand definitions in AR model you're gonna have a bad time.

Foreign keys and relationships is a tough problem and a lot of though needs to be put in in order to define a db schema with all the update/delete rules that make sense. In my opinion, having the aggregate root table (the "god" model) have cascade rules makes a lot of sense, and being frightened of this probably means that you might have an inapropriate relatinship with an inapropriate on_delete rule. As you said, it feels annoying that you get errors and have to delete relationships manually of an aggregate root model before you can delete it.

Just imagine wanting to get rid of a car but in order to do that you would have to take apart every piece of it and get rid of that thing first. Annoying, but it some cases it makes sense (when you have to trash the car's tyres somewhere else than the whole car itself)

TL;DR:
Dependant rules for associations are a business thing and I believe its usually best that both layers of the application (AR and DB) implement the same rules in the same way.

Comment (cilim) [reply db/migrate/20200910094305_create_bookings.rb:4]: >  I believe its usually best that both layers of the application (AR and DB) implement the same rules in the same way.

Absolutely agree with this! Don't know if I sounded like my intention is that the db level and app level aren't in synch regarding this behaviour.

What I mean was:
**db**
`foreign_key: true`

**app**
`dependend: :restrict_with_error`

**db**
`foreign_key: { on_delete: :cascade }`

**app**
`dependend: :destroy`

> Dependant rules for associations are a business thing

Also agree on this, and I think this question should be answered by the client (or discuss it before implementing) and not follow 1 rule strictly all the time on all projects.

File: app/models/user.rb (hunk 2 above)
Comment (lovro-bikic) [reply app/models/user.rb:19]: I think validating email address with regex is a bit pointless, since we tend to make assumptions about what is and what isn't a valid address, when in reality email addresses vary wildly, and it's frustrating for users when we block their "invalid" address because it contains a special character, or because the domain looks like a typo, or something else trivial which we assume to be invalid when in fact it isn't. This article makes a good case: https://davidcel.is/posts/stop-validating-email-addresses-with-regex/.

To drive the point home, the regex above fails on this address from the article `"Look at all these spaces!"@example.com`, which is a valid email address according to the RFC.

I think that email addresses should be validated on the client-side and if they fail some checks, it should be presented as a *warning*, not as an error, to the user, since even those libraries make mistakes and the email could still be valid. But the only viable solution on backend for email address validation is to send an email and check whether it comes through.

Another article on the topic with more examples of valid quirky emails: https://haacked.com/archive/2007/08/21/i-knew-how-to-validate-an-email-address-until-i.aspx/

Comment (jerko-culina) [reply app/models/user.rb:19]: I would say that we should check email addresses on the backend and the easiest way is with regex. Everything that we need to do is to check that we have at least 2(or 3) characters followed by @ followed by al least 2 characters followed by dot followed by at least 1 character.

Comment (JureCindro) [reply app/models/user.rb:19]: @lovro-bikic 
What about easing off the first part of the regex, to something like `[^@]+`. I think the point of this email regex validation is to have a validation that still helps the user find typos while keeping the validation cost low (in comparison to validating emails with mailgun's api or something similar)

@jerko-culina 
1. why 2 or 3 characters before `@`?
2. why 2 characters before the TLD?

You've actually suggested a way more restrictive regex than the "complicated", albeit a bit verbose, one above, which allows `a@a.a`, `a@a-a.a`  but not `a@-.a` or `a@a-.a` or `a@a.-`

Comment (uncoverd) [reply app/models/user.rb:19]: What we consider a valid email does not match what other services consider a valid email. Mailgun has their own regex for example where they check for current TLDs and more. We have instances where the email regex check passes and Mailgun won't send the actual email since it "looks funny".

 I agree with what @lovro-bikic said, a relaxed regex + warning (instead of an error) would be the best solution here

Comment (jerko-culina) [reply app/models/user.rb:19]: >why 2 or 3 characters before @?
why 2 characters before the TLD?

At the end, I added (or pattern similar to that), I used it just as a way to express what I wanted to say.

>EMAIL_REGEXP = %r{
    \A
    [a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@
    [a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?
    (?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+
    \z
  }x.freeze

This one is not complex it is complicated.

An approach like: `[^@]+` (you just suggested) is simple.

Comment (jerko-culina) [reply app/models/user.rb:19]: > I agree with what @lovro-bikic said, a relaxed regex + warning (instead of an error) would be the best solution here

Comment (JureCindro) [reply app/models/user.rb:19]: > At the end, I added (or pattern similar to that), I used it just as a way to express what I wanted to say.
> 
> > EMAIL_REGEXP = %r{
> > \A
> > [a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@
> > a-zA-Z0-9?
> > (?:.a-zA-Z0-9?)+
> > \z
> > }x.freeze

Never mind for not answering my question, it does not actually matter at this point.

But I feel like I have a responsibility to say this: I hope you're not using the regex you typed in production code. I'm hoping you just copied the regex cilim mentioned and "simplified" it without actually testing, in a hurried attempt to prove a point. 
Because it is invalid and will reject most emails except subdomains and domains domains that are literally `a-zA-Z0-9`, with top level domains that are literally `a-zA-Z0-9` (example: `jure@a-zA-Z0-9.a-zA-Z0-9` vould pass but `jure@infinum.co` would not)

Comment (jerko-culina) [reply app/models/user.rb:19]: >But I feel like I have a responsibility to say this: I hope you're not using the regex you typed in production code. I'm hoping you just copied the regex cilim mentioned and "simplified" it without actually testing, in a hurried attempt to prove a point.
Because it is invalid and will reject most emails except subdomains and domains domains that are literally a-zA-Z0-9, with top level domains that are literally a-zA-Z0-9 (example: jure@a-zA-Z0-9.a-zA-Z0-9 vould pass but jure@infinum.co would not)

Just as info, I am a little bit slow but I know read regex.

Also about not answering to your question: 
I wrote:

Comment (lovro-bikic) [reply app/models/user.rb:19]: > @lovro-bikic What about easing off the first part of the regex, to something like [^@]+.

@JureCindro that seems reasonable until a user comes with this email `"why.im.i.not.v@lid"@domain.com`. You can check that it's valid [here](https://isemail.info/%22why.im.i.not.v%40lid%22%40domain.com).

From the [RFC](https://tools.ietf.org/html/rfc5322#section-3.4.1):
> An addr-spec is a specific Internet identifier that contains a locally interpreted string followed by the at-sign character ("@", ASCII value 64) followed by an Internet domain.  The locally interpreted string is either a quoted-string or a dot-atom.

We're used to the "normal" pre-@ strings which the RFC calls dot-atoms (`lovro.bikic`, `jure.cindro`, etc.), however it also supports quoted-strings inside of which you can use pretty much anything, including `@`, so `"lovro@bikic"` is also valid and doesn't break the rules.

I think the only assumption you can make about email address format is: `it's not an email address if it doesn't contain a @`. With that in mind, this regex from the blog post above makes most sense to me: `/.+@.+\..+/i`. It checks that there is something in front of `@` and something after the `@` with a dot somewhere. (Even that breaks the rules since [this](https://isemail.info/lovro%40localhost) should be accepted.)

All other assumptions (like min 2-3 characters before and after `@`) don't hold, and if we make them, the future is bound to bite us in the ass. (`l@bikic.com` is valid and it can exist.)

This has become a really long thread, but I hope that all the comments and tries to come up with an ultimate pattern for emails have shown us how fruitless the undertaking actually is. If it sends, it's an email. 🙂

Comment (melcha) [reply app/models/user.rb:19]: C/p from devise config:

```ruby
  # Email regex used to validate email formats. It simply asserts that
  # one (and only one) @ exists in the given string. This is mainly
  # to give user feedback and not to assert the e-mail validity.
  config.email_regexp = /\A[^@\s]+@[^@\s]+\z/
```

And this is enabled by default, if you include `validatable` module. 

In all apps I add user model through devise, I thought most of us do. Why are we adding user model here manually?

File: db/migrate/20200910094305_create_bookings.rb (hunk 7 above)
Comment (melcha) [reply db/migrate/20200910094305_create_bookings.rb:7]: Let's not complicate it too much on purpose for now, but we can add this later if needed, if it will help with displaying some coding practices that are not yet present in the app.

File: db/migrate/20200910094305_create_bookings.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+      t.integer :seat_price
```
Comment (melcha) [inline db/migrate/20200910094305_create_bookings.rb:8]: My question is: is a booking (or any other model) valid without the seat_price, no_of_seats? 
If not, I'd add `null: false` on all required fields.

File: db/schema.rb [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    t.integer "no_of_seats"
+    t.integer "seat_price"
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+    t.index ["flight_id"], name: "index_bookings_on_flight_id"
+    t.index ["user_id"], name: "index_bookings_on_user_id"
+  end
+
+  create_table "companies", force: :cascade do |t|
+    t.string "name", null: false
```
Comment (melcha) [inline db/schema.rb:30]: This one should also be unique.

File: app/models/user.rb (hunk 2 above)
Comment (martincekada) [reply app/models/user.rb:19]: I will add devise in next PR (with first controller that needs authentication). I can not recall why didnt I use it in the first place, probably because I wanted to keep PRs as simple as possible (because even small PRs create good discussions which I love and dont want make them to big so that people dont have time/energy for discussions).

Have we reached consensus on email regex format? Is devises default good enough for us?

File: db/migrate/20200910094305_create_bookings.rb (hunk 7 above)
Comment (martincekada) [reply db/migrate/20200910094305_create_bookings.rb:7]: Agree with @melcha.
My approach to this app is that I look for coding practices we want to display. Then I try to think of feature that makes sense where I could display that coding practice. The bigger the pool of feature ideas is, it will be easier for me to create feature that make sense and display coding practice so I am welcoming all ideas. But I dont think it is efficient to implement features just so that app would look more like real world app if it would not bring anything new to the table.

So if you have any feature/coding practice idea, dont hesitate to suggest in any channel of communication.

File: db/migrate/20200910094305_create_bookings.rb (hunk 8 above)
Comment (martincekada) [reply db/migrate/20200910094305_create_bookings.rb:8]: Agree, will correct!

File: db/schema.rb (hunk 9 above)
Comment (martincekada) [reply db/schema.rb:30]: Agree

**Comment by martincekada** [comment]
I believe I pushed all requested changes, so please comment if there is anything else you think should be different.

Only non consensual thing in this PR was email validation format. I will definitely link somewhere this whole discussion, just have to decide for the right place (comment in code, coding_agreement, readme, doc?). But for now, I would go with devices default email format.

**Review by melcha — APPROVED** [review]

**Review by lovro-bikic — APPROVED** [review]

**Review by d4be4st — APPROVED** [review]

File: db/migrate/20200910094305_create_bookings.rb (hunk 5 above)
Comment (vr4b4c) [reply db/migrate/20200910094305_create_bookings.rb:4]: > My gripe with using this is as follows:
If there was no other choice, and I was brave enough to go editing records in the DB directly, I would forget about this option and delete records I didn't want to accidentally. We're used to the concept of dependent: :destroy in Rails, but I don't think we carry that thought over to SQL.

_and I was brave enough to go editing records in the DB directly_ 
this is a *bad* starting point for any hypothesis (supporting or opposing)
- direct DB edits are very rare (or at least they should be)
- it's extremely difficult to account for every disaster that could happen during direct DB edit (so where do we draw the line?) 

> Some articles (e.g. https://thoughtbot.com/blog/referential-integrity-with-foreign-keys) seem to even suggest using just cascades and ditching the callbacks.

If there is no complex (eg. sending email) logic that needs to be performed on delete, then DB level delete is perfectly feasible approach.

> You should use AR's destroy method most of the times which uses the dependent option and destroys the related records, sql foreign key cascade wont run in those cases as AR destoys the records before the parent record is deleted.

So you are saying (and then I agree with you!) having both `on_delete: :cascade` and `dependent: :destroy` doesn't make sense because cascade will never be triggered (normal circumstances)?
cc @JureCindro 

> Dependant rules for associations are a business thing and I believe its usually best that both layers of the application (AR and DB) implement the same rules in the same way.

And then concluding we should keep them in sync, 😕   what???

> Absolutely agree with this! Don't know if I sounded like my intention is that the db level and app level aren't in synch 
regarding this behaviour.
> db
> foreign_key: { on_delete: :cascade }
> app
> dependend: :destroy

🆙    cc @cilim 

cc @martincekada

Comment (JureCindro) [reply db/migrate/20200910094305_create_bookings.rb:4]: > > You should use AR's destroy method most of the times which uses the dependent option and destroys the related records, sql foreign key cascade wont run in those cases as AR destoys the records before the parent record is deleted.
> 
> So you are saying (and then I agree with you!) having both `on_delete: :cascade` and `dependent: :destroy` doesn't make sense because cascade will never be triggered (normal circumstances)?
> cc @JureCindro
> 
> > Dependant rules for associations are a business thing and I believe its usually best that both layers of the application (AR and DB) implement the same rules in the same way.
> 
> And then concluding we should keep them in sync, 😕 what???

Well the db constraints are often there for the not so normal circumstances. 

My personal preference is keeping them in sync. I've had situations where I needed to truncate some tables on staging and it felt great that I could do that with `delete`/`delete_all` instead of `destroy`. 
The cascade rules can reveal a lot about data when looking at the database diagram, so having that there might help.

Comment (vr4b4c) [reply db/migrate/20200910094305_create_bookings.rb:4]: >  I’ve had situations where I needed to truncate some tables on staging and it felt great that I could do that with delete/delete_all instead of destroy.

Cascading delete can be both blessing and disaster. Use the following questionnaire to determine which one is it.
After successful deletion of a record
• are associated records also deleted?
• should they be deleted?

Comment (JureCindro) [reply db/migrate/20200910094305_create_bookings.rb:4]: > > I’ve had situations where I needed to truncate some tables on staging and it felt great that I could do that with delete/delete_all instead of destroy.
> 
> Cascading delete can be both blessing and disaster. Use the following questionnaire to determine which one is it.
> After successful deletion of a record
> • are associated records also deleted?
> • should they be deleted?

That questionnaire applies to both cascade delete and dependant options in AR relationships 😊

Comment (vr4b4c) [reply db/migrate/20200910094305_create_bookings.rb:4]: > That questionnaire applies to both cascade delete and dependant options in AR relationships 😊

Yep, I totally agree. I was going in the direction, if you are doing something on production, come prepared and informed, otherwise, you are in a lottery where the winning ticket is coming out unharmed and all others are a disaster.

And if anyone thinks, _every little bit of safety matters_, I'm saying no it doesn't! If you are messing with production, come prepared and informed, apply changes with surgical precision, get in and out as fast as you can touching as little as possible. 

To all other arguments about `prod-console-safety`, that is an illusion, and if you believe it I'm volunteering to pretend a misinformed, bad decision-making developer with prod console access. Lunch is on me if I don't cause a disaster in the first 2 minutes.

> I’ve had situations where I needed to truncate some tables on staging and it felt great that I could do that with delete/delete_all instead of destroy.

OK, I have some experience there
- if you have one or two level of children to dispose, raw sql isn't that dificult
- there is also option of using `TRUNCATE table_name CASCADE;`
- anything more complex than that, refer to the begining of my comment

> The cascade rules can reveal a lot about data when looking at the database diagram, so having that there might help.

What information does it reveal other than, this record will be deleted/restricted/nullified/nothing when parent record (fk) is deleted?
When you say db diagram what do you mean?
 - `.png export` - I have never seen and export which shows cascade option
 - `db diagram tool (eg. MySqlWorkbench, sqldbm.com, dbdiagram.io, ...)` - either not suppored or `double-click table` -> `foreign keys/associations tab` -> `fk/assoc details` (not very visible)
 - something else?

I find it most visible in Postico, but even that is far from obvious.

For all intents and purposes, `cascade` and `dependent-destroy` are orthogonal tools, you either
- need one
- nedd other
- you don't care and can use which ever you want

*cascade*
- in situations when you don't need to act on delete event, or
- you don't need to act on the delete event and you need query performance

*dependent-destroy*
- when you need to act on the delete event
- [subjective] it's a good default (optimize with cascade where necessary)

I see a potential for debate about primary option, but since I don't have a strong opinion on either one of them, I'm perfectly fine using both of them, interchangeably. Personaly I prefer the `dependent-destroy` because I consider ORM as primary tool for DB interaction and in that context it feels natural.

Comment (JureCindro) [reply db/migrate/20200910094305_create_bookings.rb:4]: Ok, I wasn't saying that they provide an *easy* way about looking at relationships, just that they might (if used properly), provide a bit of insight on what the aggregate roots of a bunch of tables are.

I still don't know why wouldn't you keep both things, even if you "always" only use AR's destroy actions. Because we are lazy and hate typing? 
I still see a benefit since you're defining the foreign key from the other side of the relationship as when defining the `dependant` option in AR, so you are forced to look at the relationship from another perspective.

We're making this sound like it is a big issue but it really isn't. But I'd be vary of using only the db cascades, we would somehow need to ensure that the model deleted by the cascade has no delete callbacks set. I have no idea what any gems that we are using add to models so that is one of the main reasons I'd default to `dependent:` options.

Comment (vr4b4c) [reply db/migrate/20200910094305_create_bookings.rb:4]: Q:
 > I still don't know why wouldn't you keep both things, even if you "always" only use AR's destroy actions. Because we are lazy and hate typing?

A:
> But I'd be vary of using only the db cascades, we would somehow need to ensure that the model deleted by the cascade has no delete callbacks set. I have no idea what any gems that we are using add to models so that is one of the main reasons I'd default to dependent: options.

Kind of sums it up (at least for me it does). In my case, I default to `dependent-destroy` (let ORM handle everything), and if there is a `cascade` defined (+ `dependent-destroy` turned OFF) it means I purposely decided it's OK for DB to handle this one (there are no delete callbacks).

So my premise is they are orthogonal and it doesn't make sense for them to coexist at the same time on the same reference (literally one is to account for the fallacies of the other). So to expand on my premise, if they are both "enabled" that means I can use both ORM and DB to delete records, which as you pointed out isn't true if you have callbacks defined.
